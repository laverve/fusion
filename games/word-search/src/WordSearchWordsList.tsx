import React, { useContext, useMemo } from "react";
import classnames from "classnames";
import { parseToRgba, toHex, readableColor } from "color2k";
import { Chip } from "@nextui-org/react";
import { WordSearchContext } from "./WordSearch.context";

export type WordSearchWordsListProps = {
    classNames?: string;
};

export const WordSearchWordsList: React.FC<WordSearchWordsListProps> = ({ classNames }: WordSearchWordsListProps) => {
    const { words, foundWords, selectedWordsColors } = useContext(WordSearchContext);

    const lowerCasedFoundWords = useMemo(() => foundWords.map((w) => w.toLowerCase()), [foundWords]);

    const isWordSelected = (word: string) => lowerCasedFoundWords.findIndex((w) => w === word.toLowerCase()) > -1;

    const selectedWordsColorsWithoutOpacity = useMemo(
        () =>
            selectedWordsColors.map((color) => {
                const [r, g, b] = parseToRgba(color);
                return toHex(`rgb(${r},${g},${b})`);
            }),
        [selectedWordsColors]
    );

    const getSelectedWordColor = (word: string) => {
        const idx = lowerCasedFoundWords.findIndex((w) => w === word.toLowerCase());
        return selectedWordsColorsWithoutOpacity[idx % selectedWordsColorsWithoutOpacity.length];
    };

    return (
        <div
            data-testid="testid-wordsearch-words-list"
            className={classnames("box-border", "justify-center", "items-center", "flex-1", "gap-1", classNames)}
        >
            {words.map((word) => (
                <Chip
                    key={word}
                    data-testid={`testid-${word.toLowerCase()}`}
                    className="m-1"
                    style={{
                        ...(selectedWordsColorsWithoutOpacity.length > 0 && isWordSelected(word)
                            ? {
                                  backgroundColor: getSelectedWordColor(word),
                                  color: readableColor(getSelectedWordColor(word))
                              }
                            : {})
                    }}
                >
                    <span className={classnames({ "line-through": isWordSelected(word) })}>{word}</span>
                </Chip>
            ))}
        </div>
    );
};
