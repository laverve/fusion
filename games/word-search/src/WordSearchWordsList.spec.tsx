/**
 * @jest-environment jsdom
 */

import { describe, it, expect } from "@jest/globals";
import React, { useContext, useState } from "react";
import snapshotDiff from "snapshot-diff";

import { render, fireEvent } from "@testing-library/react";

import { WordSearchContextProvider, WordSearchContext } from "./WordSearch.context";
import { WordSearchWordsList, WordSearchWordsListProps } from "./WordSearchWordsList";

export const TestOnWordFound: React.FC<{ words: string[] }> = ({ words }) => {
    const { onWordFound } = useContext(WordSearchContext);
    const [idx, setIdx] = useState(0);
    return (
        <button
            type="button"
            data-testid="test-on-word-found"
            onClick={() => {
                onWordFound(words[idx]);
                setIdx(idx + 1);
            }}
        >
            Find word
        </button>
    );
};

describe("WordSearchWordsList", () => {
    const renderComponentUderTest = ({
        props = {},
        selectedWords = ["cat"],
        selectedWordsColors = ["black"]
    }: {
        props?: WordSearchWordsListProps;
        selectedWords?: string[];
        selectedWordsColors?: string[];
    } = {}) => {
        const words = ["dog", "cat", "dig", "boat", "doc"];
        const grid = [
            ["b", "d", "i", "g"],
            ["d", "o", "g", "o"],
            ["u", "c", "a", "t"],
            ["x", "y", "z", "t"]
        ];

        return render(
            <WordSearchContextProvider grid={grid} words={words} selectedWordsColors={selectedWordsColors}>
                <TestOnWordFound words={selectedWords} />
                <WordSearchWordsList {...props} />
            </WordSearchContextProvider>
        );
    };

    it("should render all words as unselected", () => {
        const component = renderComponentUderTest();
        expect(component.asFragment()).toMatchSnapshot();
    });

    it("should render selected words as selected", () => {
        const component = renderComponentUderTest({ selectedWords: ["cat"] });
        const initialFragment = component.asFragment();
        fireEvent.click(component.getByTestId("test-on-word-found"));
        expect(snapshotDiff(initialFragment, component.asFragment())).toMatchSnapshot();
    });

    it("should use color without alpha channel to hightlight selected word", () => {
        const component = renderComponentUderTest({ selectedWords: ["cat"], selectedWordsColors: ["rgba(0,0,0,0.5)"] });
        const initialFragment = component.asFragment();
        fireEvent.click(component.getByTestId("test-on-word-found"));
        expect(snapshotDiff(initialFragment, component.asFragment())).toMatchSnapshot();
    });

    it("should use colors from the list in the same order as words were found", () => {
        const component = renderComponentUderTest({
            selectedWords: ["dig", "cat", "dog"],
            selectedWordsColors: ["rgba(0,0,0,0.5)", "rgba(1,1,1,0.5)", "rgba(2,2,2,0.5)"]
        });
        fireEvent.click(component.getByTestId("test-on-word-found"));
        fireEvent.click(component.getByTestId("test-on-word-found"));
        fireEvent.click(component.getByTestId("test-on-word-found"));
        expect(component.getByTestId("testid-dig").style.backgroundColor).toBe("rgb(0, 0, 0)");
        expect(component.getByTestId("testid-cat").style.backgroundColor).toBe("rgb(1, 1, 1)");
        expect(component.getByTestId("testid-dog").style.backgroundColor).toBe("rgb(2, 2, 2)");
    });
});
