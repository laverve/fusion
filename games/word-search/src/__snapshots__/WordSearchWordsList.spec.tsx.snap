// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordSearchWordsList should render all words as unselected 1`] = `
<DocumentFragment>
  <div
    data-overlay-container="true"
  >
    <button
      data-testid="test-on-word-found"
      type="button"
    >
      Find word
    </button>
    <div
      class="box-border justify-center items-center flex-1 gap-1"
      data-testid="testid-wordsearch-words-list"
    >
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
        data-testid="testid-dog"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          <span
            class=""
          >
            dog
          </span>
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
        data-testid="testid-cat"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          <span
            class=""
          >
            cat
          </span>
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
        data-testid="testid-dig"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          <span
            class=""
          >
            dig
          </span>
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
        data-testid="testid-boat"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          <span
            class=""
          >
            boat
          </span>
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
        data-testid="testid-doc"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          <span
            class=""
          >
            doc
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`WordSearchWordsList should render selected words as selected 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -27,16 +27,17 @@
          </span>
        </div>
        <div
          class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
          data-testid="testid-cat"
+         style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          <span
            class="flex-1 text-inherit font-normal px-2"
          >
            <span
-             class=""
+             class="line-through"
            >
              cat
            </span>
          </span>
        </div>"
`;

exports[`WordSearchWordsList should use color without alpha channel to hightlight selected word 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -27,16 +27,17 @@
          </span>
        </div>
        <div
          class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground m-1"
          data-testid="testid-cat"
+         style="background-color: rgb(0, 0, 0); color: rgb(255, 255, 255);"
        >
          <span
            class="flex-1 text-inherit font-normal px-2"
          >
            <span
-             class=""
+             class="line-through"
            >
              cat
            </span>
          </span>
        </div>"
`;
